### ISSUE TITLE
Kick (or ban) users from protected rooms that are on the deny ACL.

### ISSUE THREAD

**Issue by Gnuxie** [matrix-org/mjolnir#161]
Users can still see the room timeline when their server is banned (that's not always wanted).
It's basically impossible to be sure ACL works anymore from the perspective of a non-synapse admin or someone who doesn't have time for manual intervention, it's only a matter of time before this is exploited and we should be ready for that. See https://matrix.to/#/!NasysSDfxKxZBzJJoE:matrix.org/$mHtLd-SlLRxqfxrMAk2zNYBYly_dD7AWLxzLo9dyRiE?via=matrix.org&via=libera.chat&via=envs.net.

**Comment by Yoric** [comment]
I'm missing a little context.

Do I understand correctly that it's a case in which:

- user @user:evil.com is already in `room`;
- server evil.com is ACL-ed away from `room`;
- @user:evil.com can still receive events in `room`;
- @user:evil.com can still send events in `room`?

**Comment by Gnuxie** [comment]
Yes, so long as there is a server that isn't implementing ACL in the room that leaks the events (in both directions, so everyone's events will get leaked to evil.com and new events from evil.com will get leaked to everyone else via the leaky server). If you kick the evil.com users then the kicks will get propagated to evil.com via the leaky server too. What you can't stop with kicks or bans is new evil.com users joining the room though (even if they are immediately kicked or banned by mjolnir).

**Comment by Gnuxie** [comment]
Should be implemented in conjunction with https://github.com/matrix-org/mjolnir/issues/279 and the option to kick by default should probably be placed behind config.

We will also want to rate limit the kicking mechanism in both cases  so that do not cause as much damage as a mass join incident. I think it should be made clear that for an intolerable mass join incident where Synapse can't keep up, this kicking is essentially useless, so explicitly warn against someone removing the rate limit to try defend against such an attack.
